server/src/hardware/protocol/z21/clientkernel.cpp

```diff
@@ -483,8 +523,18 @@ void ClientKernel::send(const Message& message)
 
 void ClientKernel::startKeepAliveTimer()
 {
-  assert(ClientConfig::keepAliveInterval > 0);
-  m_keepAliveTimer.expires_after(boost::asio::chrono::seconds(ClientConfig::keepAliveInterval));
+  if(m_broadcastFlags == BroadcastFlags::None)
```

Comment:
Shoudn't we test against the `requiredBroadcastFlags` ?

Reply:
> Shoudn't we test against the `requiredBroadcastFlags` ?

For the arduino Z21 implementation, some flags are not supported (e.g. CAN Bus) and some other are forced to be enabled together (e.g. enabling something about LocoNet, it enables every LocoNet flag).
So in general there is a negotiation on the flag value, it may be more than what was requested or some features may be missing.

Reply:
I think it would be good to log a warning if there is a mismatch or the retry count expires.
Then the user at least knows something might not work as expected.

You can add two warnings for that in the 2xxx range.

Reply:
For now I've added only one message...

Two things:
- Do you think it's important also to log the effective returned flags?
- Also do we consider a mismatch if reply has more flags than requested?
Technically it's still a mismatch, but we receive all things requested and just ignore extra broadcasted messages we are not interested in.
We can test it like: `(flags & requiredFlags) == requiredFlags`

Reply:
> Do you think it's important also to log the effective returned flags?

I'm a bit in doubt about it, for us as developers it it useful information, however for an average user it is useless. If e.g. someone in the future runs into it, we can just instruct them to enable debug logging and send the log. That will give us as developers much more information to work with. So my vote is for don't log the flags in the message. What is your opinion?

> Also do we consider a mismatch if reply has more flags than requested?
Technically it's still a mismatch, but we receive all things requested and just ignore extra broadcasted messages we are not interested in.

Good point, It is still a mismatch, Traintastic can behave as expected so no need to bother the user with it. We could how log it as extra debug message if debug logging is enabled (makes it a bit easier to spot for us as developers).
What is your opinion?

Reply:
I agree with you. So we will have 2 messages:
- A warning emitted only if BC flags are less than requested
- A debug (hidden by default) emitted always and with flag value

Reply:
Yes indeed :)